Support configurable delimiter for seed files, default to comma (#3990)

ramonvermeulen opened the pull request:
resolves #3990

### Description

Added support to configure a custom delimiter for seed files, defaults to comma.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR -> https://github.com/dbt-labs/docs.getdbt.com/issues/3015
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

jtcohen6 commented on core/dbt/contracts/graph/model_config.py:
```diff
@@ -435,6 +435,7 @@ class NodeConfig(NodeAndTestConfig):
         default_factory=dict,
         metadata=MergeBehavior.Update.meta(),
     )
+    delimiter: str = ","
```
This should go in [`SeedConfig`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/contracts/graph/model_config.py#L500), now defined separately, rather than `NodeConfig`

jtcohen6 commented on core/dbt/clients/agate_helper.py:
```diff
@@ -135,12 +135,12 @@ def as_matrix(table):
 
 
-def from_csv(abspath, text_columns):
+def from_csv(abspath, text_columns, delimiter = ","):
```
**Question:** Should we add support for more than just `delimiter` here? Any/all `**kwargs` supported by `agate.from_csv` / Python's stdlib CSV reader?

**Thoughts:**
- That can be out of scope for now / for this PR
- It would risk locking us further into using `agate` (which I don't love), or at least the Python stdlib `csv` reader (which I don't mind as much)

JCZuurmond replied:
```suggestion
def from_csv(abspath, text_columns, delimiter=","):
```

jtcohen6 reviewed (commented):
Thanks @ramonvermeulen! I think we'd want a functional test for this as well, guaranteeing that a CSV with a nonstandard (non-comma) delimiter actually works.

From my POV, this doesn't feel super high-priority — dbt's `seed` functionality is intentionally limited / narrow in scope, and I hesitate before adding more, given the added maintenance burden — but this is a very straightforward & self-contained change. TSVs are common enough in the wild.

ramonvermeulen commented:
> Thanks @ramonvermeulen! I think we'd want a functional test for this as well, guaranteeing that a CSV with a nonstandard (non-comma) delimiter actually works.
> 
> From my POV, this doesn't feel super high-priority — dbt's `seed` functionality is intentionally limited / narrow in scope, and I hesitate before adding more, given the added maintenance burden — but this is a very straightforward & self-contained change. TSVs are common enough in the wild.

Hi @jtcohen6, sorry for the late reply, I had a week of vacation.
I think adding a functional test is a good idea! I will dive into the functional test set-up and see if I can make a good functional test.

The reason I made this PR is because I encountered customers delivering there "csv" files with for example pipes or tabs as delimiter. Maybe because that is a standard within their organization. I thought it would be nice to have some more control about the configuration of the csv parser for the seed files, but I get your point about that it will increase the maintenance. 


> **Question:** Should we add support for more than just `delimiter` here? Any/all `**kwargs` supported by `agate.from_csv` / Python's stdlib CSV reader?
> 
> **Thoughts:**
> 
> * That can be out of scope for now / for this PR
> * It would risk locking us further into using `agate` (which I don't love), or at least the Python stdlib `csv` reader (which I don't mind as much)

I find this a hard one, it is possible to add support for all the params of the `agate.from_csv` function, however this would lock dbt further in to using `agate`. The `delimiter` argument, is an argument that is also [supported](https://docs.python.org/3/library/csv.html#dialects-and-formatting-parameters) by the stdlib `csv` reader. I would say support for other arguments is out of scope for now.

JCZuurmond reviewed (approved):
LGTM, @ramonvermeulen!

eliwoods commented:
@ramonvermeulen I am rather interested in this PR, is there anything I can do to help needed to get it over the line?

ramonvermeulen commented:
> @ramonvermeulen I am rather interested in this PR, is there anything I can do to help get it over the line?

Hi @eliwoods, I think functional tests are still required to get this PR over the line @jtcohen6?  I've been quite busy lately so feel free to pick this up / contribute on this as you wish!

ramonvermeulen commented:
@jtcohen6 @eliwoods @JCZuurmond I added two functional tests for this feature, and fixed the failing seed related integration tests. Only one integration test called `test_snapshot_hard_delete` is failing on my local system. However this test is failing for me on the main branch as well, so I don't think it is related to this pull request.

thomas-vl reviewed (approved):
LGTM 🚀

QMalcolm reviewed (approved):
@ramonvermeulen Thank you for putting all this work it! I've looked it over and pulled it down for some manual testing. Every looks great ✨ Gonna run it through our CI workflow, and then we should be good to go 🙂

QMalcolm commented:
Ek, foiled by [mypy](https://github.com/dbt-labs/dbt-core/actions/runs/5356319599/jobs/10148851050?pr=7186). Perhaps I should have run the `pre-commit` hooks locally while reviewing 🙃 

So there are two classes of mypy issues we're seeing. One is some simple formatting changes, which are simple fixes. The other is the errors like `core/dbt/cli/main.py:403: error: <nothing> has no attribute "command"  [attr-defined]`. These errors are actually due to a recently released version of `click`, specifically `click 8.1.4`. We've [fixed this in core](https://github.com/dbt-labs/dbt-core/pull/8050), thus to fix this I'd recommend merging main into this branch to get it up to date.

ramonvermeulen commented:
> Eek, foiled by [mypy](https://github.com/dbt-labs/dbt-core/actions/runs/5356319599/jobs/10148851050?pr=7186). Perhaps I should have run the `pre-commit` hooks locally while reviewing upside_down_face
> 
> So there are two classes of mypy issues we're seeing. One is some simple formatting changes, which are simple fixes. The other is the errors like `core/dbt/cli/main.py:403: error: <nothing> has no attribute "command" [attr-defined]`. These errors are actually due to a recently released version of `click`, specifically `click 8.1.4`. We've [fixed this in core](https://github.com/dbt-labs/dbt-core/pull/8050), thus to fix this I'd recommend merging main into this branch to get it up to date.

Thanks for taking a look at my PR! Just pushed a commit merging main back into my branch, and ran pre-commit, so hopefully that fixes the CI.

QMalcolm commented:
> Thanks for taking a look at my PR! Just pushed a commit merging the upstream `main` back into my branch, and ran pre-commit, so hopefully that fixes the CI.

Absolutely! Thank you for the quick turn around with the fixes. Also, sorry that it took me so long to get to my initial review. It's been a busier few weeks for me than I anticipated 😅 I'm gonna kick off the CI again

QMalcolm reviewed (approved):
Re-approving 🙂

saviorand commented:
Installed the pre-release version and trying this in my `dbt_project.yml`, doesn't seem to work.. Not sure what's the issue:
```
seeds:
  <project_name>:
    <seed_directory_name>:
      +delimiter: ";"
```

ramonvermeulen commented:
> Installed the pre-release version and trying this in my `dbt_project.yml`, doesn't seem to work.. Not sure what's the issue:
> ```
> seeds:
>   <project_name>:
>     <seed_directory_name>:
>       +delimiter: ";"
> ```

Hmmm, not sure what is going wrong. Currently I am leaving for holidays for a week, but will definitely look into this once I am back.

CarMoreno commented:
Great feature!!. Just one point: What about TSV files? In that case, the delimiter would be `\t` and won't work. I know it can be considered an edge case, I had to change my TSV file to use `|` as the delimiter character and my seed worked.
